### ISSUE TITLE
Yolov5n on TFServing

### ISSUE THREAD

**Issue by JrGxllxgo** [ultralytics/yolov5#11584]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello, I trained with a custom dataset a model on Ultralytics Google Colab for object detection. I exported the model as saved_model fotrmat to upload to my TFServing docker, when I do the request to the docker the return I get is an array of +25000 lines, it seems like a numpy array but I´m not sure.

It is supposed to return an array whith the information of what the model found.

What can I do for getting this array and not that big array? Is there any problem to upload the yolov5n custom model on Tensor Flow Serving?

### Additional

![Captura de pantalla 2023-05-24 091119](https://github.com/ultralytics/yolov5/assets/106581652/3ae2c557-dc6a-4a51-ba68-1a88118a3db4)

**Comment by glenn-jocher** [comment]
@JrGxllxgo hello! 

It sounds like you're experiencing some issues getting your yolov5n custom model to work with Tensor Flow Serving. The large array you're seeing might be related to how the model is processing the data. To help you better, could you provide more details and information on how you are making the request to the TFServing Docker, as well as the payload (input data) you are sending? 

Also, which version of the Ultralytics YOLOv5 were you using for training the model? 

Please let us know so we can provide further assistance.

**Comment by JrGxllxgo** [comment]
@glenn-jocher thks for help me, I´ll provide information, I hope you find it useful.
All codes, enviroments... are from the google Collab of Ultralytics

- Yolov5: Yolov5n
- The input I send to the TFServing is as normal models, a JSON with a numpy array with the normal format
![request_inputdata](https://github.com/ultralytics/yolov5/assets/106581652/637c4b6f-0ef6-4d2d-88df-3f2f7897a41e)

- The url request is http://{ip}:{port}/v1/models/model01:predict

![train_output](https://github.com/ultralytics/yolov5/assets/106581652/bca46795-1c1a-49f1-a1eb-315608585b88)
![export_command](https://github.com/ultralytics/yolov5/assets/106581652/cddb4235-7c4e-45b0-bf88-b842715bec0b)
![1st step](https://github.com/ultralytics/yolov5/assets/106581652/942e236f-6655-4dd2-8636-e2db00f603d5)
![train_command](https://github.com/ultralytics/yolov5/assets/106581652/2cd43280-f4db-4cd9-ab5f-2d0f60cf0b53)

**Comment by glenn-jocher** [comment]
Hi @JrGxllxgo, thank you for providing all of that helpful information. It looks like everything is set up correctly, and it's hard to say for certain without more information, but I may have a few potential solutions. 

Firstly, it may be that the large numpy array you're seeing is the correct output from the model. Specifically, this could be what is known as a raw output of the detection. Depending on what you are trying to do, it may be necessary to do some additional post-processing and filtering of this output to get the results you're after.

Another potential solution could be to check if the model was saved successfully as a `SavedModel` format and is compatible with TFServing. You can verify this by checking if you can load the model back and run inference on it in your local environment.

Finally, if none of the solutions work, I'd recommend checking out the TFServing documentation and seeing if there are any potential issues specific to that part of your workflow. Hope this helps, let me know if you have any further questions!

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher , thks you for giving me some solutions.

If I get this result when I do a request it means that the model loaded correctly, right?
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/25e5ac64-8f65-4c93-97c9-6478eaa572b0)

During this morning I´ll try to do some post-processing or search if I the docker with TFServing should have a different build for object_detection.

**Comment by glenn-jocher** [comment]
Hi @JrGxllxgo, you're welcome! Glad to be of help.

Regarding your question, yes, it seems like the model has loaded correctly. The output you are seeing could be the raw output of the detection, and it may require further post-processing and filtering to get the results you're after.

I hope you're able to make some progress with implementing the post-processing techniques. If you continue to have issues, it may be worth looking into whether the Docker container with TFServing requires a different build for object detection.

Best of luck, and feel free to reach out if you have any more questions.

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher again!

Which parameters I should change on the training to get the typical JSON with the predictions with accuracy and coords?

I think it can be useful to resolve the problem or another way to face the problem because y post-processed the image and is 25200x1 px

**Comment by glenn-jocher** [comment]
Hi @JrGxllxgo, great question! To get the typical JSON output with the predictions including accuracy and coordinates, you will need to enable the `--save-json` flag during training and export. Here's an example training command to include this flag:

```
python train.py --img 640 --batch 16 --epochs 30 --data data.yaml --cfg models/yolov5n.yaml --weights '' --save-json --cache
```

Additionally, during export, you'll need to include the `--dynamic` flag to ensure that the model is exported with dynamic shapes that support variable batch sizes. Here's an example export command:

```
python export.py --weights runs/train/exp/weights/best.pt --include=trt --dynamic --verbose
```

Please note that you'll need to specify your own YAML file in the `--data` flag and YOLOv5 model configuration in the `--cfg` flag based on your specific project.

I hope this helps! Let me know if you have any other questions.

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher thks for that option!

If I want to train but locally with a docker but I can´t use NVIDIA GPU, just CPU, which docker image and Notebook you recommend to me??

**Comment by glenn-jocher** [comment]
Hello @JrGxllxgo, you're welcome! 

If you want to train the YOLOv5 model locally on your CPU using Docker, you can use the `ultralytics/yolov5` Docker image. Here's an example Docker command to run the image:

```
docker run --gpus all -it --rm -v "$(pwd)":/app ultralytics/yolov5:latest python train.py --img 640 --batch 16 --epochs 50 --data /app/data.yaml --cfg /app/models/yolov5s.yaml --weights ''
```

However, since you do not have an NVIDIA GPU, you'll need to modify the Docker command to use the CPU only. Here's an example of how to modify the command to run YOLOv5 on CPU only:

```
docker run -it --rm -v "$(pwd)":/app ultralytics/yolov5:latest python train.py --img 640 --batch 16 --epochs 50 --data /app/data.yaml --cfg /app/models/yolov5s.yaml --weights '' --device cpu
```

This will train the model with YOLOv5s configuration for 50 epochs on CPU only. Please note that without a GPU, the training may take significantly longer, so you may want to experiment with smaller batch sizes and fewer epochs to keep the training time manageable.

I hope this helps! Let me know if you have any other questions.

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher let's try this.

The docker should have Python>=3.7 version and PyTorch>=1.7, isn´t it?

I'll ask you the last question (I hope), can all YolovX be exported as saved_model format? I saw yesterday the UltralitycsHUB website and in the Yolo models there was a section to see how to export them and not all could be exported as saved_model.

Thks u for all!!

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/d6e54498-12c1-4418-9a5c-15dccc1fa657)

**Comment by glenn-jocher** [comment]
Hi @JrGxllxgo,

Yes, you are correct. The Docker image used for training with YOLOv5 should have Python version >= 3.7 and PyTorch version >= 1.7. You can find more information about the requirements on the YOLOv5 documentation.

Regarding your question about exporting YOLOvX to the SavedModel format, not all YOLO models can be exported in the SavedModel format. As of now, only YOLOv5 models can be exported in the SavedModel format using the export.py script provided in the official Ultralytics' YOLOv5 repository. However, other models can be exported in different formats like the ONNX format, which can be loaded and used in different frameworks.

I hope this helps! Don't hesitate to ask if you have any further questions.

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher ! I started the training locally with docker, thks u a lot!

This error is caused by the training without GPU?

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/3ef81960-bbb2-41c0-bfbd-f9c5ff845554)

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher ! Finally I'm doing on an anaconda environment with all the requirements, pytorch and python. (with cpu)

I got this error with the command

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/89a4f6e0-e7e3-4489-a450-997c1a867228)

**Comment by glenn-jocher** [comment]
@JrGxllxgo hello! It's great to hear that you're making progress with your training. Regarding the error you're encountering, it seems like the error message is indicating that it cannot import the `COCOEvaluator` class from the `detectron2.evaluation` package. 

This can be caused if you are using an incorrect version of the Detectron2 library, which may not have included this class. One potential solution is to try reinstalling the Detectron2 library and ensuring that you are using a version that includes this class.

You can try running this command to reinstall the correct version of Detectron2:

```
pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.9/index.html
```

This command installs the CPU version of Detectron2 version 0.5 (which is compatible with PyTorch 1.9), and may resolve the issue you are encountering. 

If this doesn't work, please let me know, and we can explore other troubleshooting steps.

**Comment by JrGxllxgo** [comment]
@glenn-jocher hello!

I tried to run the command and here is the error I get

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/a4d09428-d89f-4ad1-8d26-261eca7653d6)

And this is the pytorch I get

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/d28b9845-32ae-4224-84da-843e4796f4f7)

**Comment by glenn-jocher** [comment]
@JrGxllxgo hello there! 

Thanks for reaching out to us.

The error seems to be caused by a version mismatch between PyTorch and Detectron2. To clarify, which version of PyTorch and which version of Detectron2 are you currently using? 

Also, can you please try running the following command to install Detectron2 and its dependencies, and see if this resolves the issue?

```
pip install -U 'git+https://github.com/facebookresearch/fvcore'
pip install -U 'git+https://github.com/cocodataset/cocoapi.git#subdirectory=PythonAPI'
pip install -U detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.7/index.html
```

Once you have installed Detectron2, please try running the command again and let us know if it works.

Thank you and please let us know if you have any additional questions or concerns.

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher 

I did your commands but I get these errors:

![imagen](https://github.com/ultralytics/yolov5/assets/106581652/e5703fee-f60a-4ba6-bc2e-911b0c532689)

So I tried this another method and I got errors
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/95b5f0b2-f3cf-4784-801f-70a5222d9570)


Doing on an anaconda env could be a problem?

**Comment by JrGxllxgo** [comment]
Hi @glenn-jocher again!

I have tried to put same commands on the Anaconda powershell as admin and these are the results y get:

pip install -U 'git+https://github.com/facebookresearch/fvcore'
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/f2bd2998-4d13-46fc-b0d1-8bca404699ce)

pip install -U 'git+https://github.com/cocodataset/cocoapi.git#subdirectory=PythonAPI'
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/6c214e35-12bb-4b20-8dd4-3d2fff3c743e)
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/ea1a3aae-171a-470f-bc2f-4f698246c646)

pip install -U detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.7/index.html
![imagen](https://github.com/ultralytics/yolov5/assets/106581652/4d38f691-70d8-4a58-8bcf-763962f3bfbd)

**Comment by glenn-jocher** [comment]
Hi @JrGxllxgo,

It looks like there are several issues when installing the required packages for Detectron2. Working within an Anaconda environment should not be a problem, but it seems that the installation commands are not completing successfully.

Let's try a different approach. First, remove any previously installed versions of Detectron2 and its dependencies, and then try the following commands:

1. Install fvcore:
```
pip uninstall -y fvcore
pip install fvcore==0.1.5.post20210605
```

2. Install pycocotools:
```
pip uninstall -y pycocotools
pip install 'git+https://github.com/cocodataset/cocoapi.git#subdirectory=PythonAPI'
```

3. Install detectron2:
```
pip uninstall -y detectron2
pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.7/index.html
```

Please run these commands one by one and carefully observe any error messages that may occur. If you encounter any issues, please provide the error messages so that we can further assist you.

Thank you for your patience, and I'm here to help with any further questions or concerns you may have.
